File: backend/app/schemas/subscription.py
"""邮局订报数据生成模块 · Pydantic schemas（In/Out 分离）。"""

from datetime import date, datetime
from decimal import Decimal
from typing import List, Optional

from pydantic import BaseModel, Field

from app.models.subscription_batch import (
    SubscriptionArtifactType,
    SubscriptionBatchStatus,
    SubscriptionImportStatus,
    SubscriptionIssueLevel,
    SubscriptionRunStatus,
)


# --- 批次 --------------------------------------------------------------------

class BatchCreateIn(BaseModel):
    year: int = Field(ge=2000, le=2100)
    start_month: int = Field(ge=1, le=12)
    make_date: Optional[date] = None
    unit_price: Optional[Decimal] = Field(default=None, ge=0)
    notes: Optional[str] = None


class BatchOut(BaseModel):
    id: int
    year: int
    start_month: int
    make_date: Optional[date] = None
    unit_price: Optional[Decimal] = None
    status: SubscriptionBatchStatus
    active_version_id: Optional[int] = None
    notes: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class BatchDetailOut(BatchOut):
    versions: List["ImportVersionOut"] = []


# --- 导入版本 ----------------------------------------------------------------

class SourceFileOut(BaseModel):
    id: int
    file_role: str
    file_type: Optional[str] = None
    original_filename: str
    size: Optional[int] = None
    sha256: Optional[str] = None

    model_config = {"from_attributes": True}


class ImportVersionOut(BaseModel):
    id: int
    batch_id: int
    version_no: int
    status: SubscriptionImportStatus
    reason: Optional[str] = None
    summary_json: Optional[dict] = None
    uploaded_at: Optional[datetime] = None
    source_files: List[SourceFileOut] = []

    model_config = {"from_attributes": True}


class ValidationIssueOut(BaseModel):
    id: int
    level: SubscriptionIssueLevel
    source: Optional[str] = None
    sheet_or_file: Optional[str] = None
    row_no: Optional[int] = None
    field: Optional[str] = None
    code: Optional[str] = None
    message: str

    model_config = {"from_attributes": True}


class ImportStatusOut(BaseModel):
    """导入版本的解析/校验状态 + 问题计数。"""

    version: ImportVersionOut
    issue_counts: dict  # {block: n, warn: n, info: n}
    can_activate: bool


# --- 生成 --------------------------------------------------------------------

class ArtifactOut(BaseModel):
    id: int
    artifact_type: SubscriptionArtifactType
    region_name: Optional[str] = None
    filename: str
    sha256: Optional[str] = None
    is_historical: bool
    created_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class GenerationRunOut(BaseModel):
    id: int
    batch_id: int
    version_id: int
    rule_version: Optional[str] = None
    template_version: Optional[str] = None
    status: SubscriptionRunStatus
    started_at: Optional[datetime] = None
    ended_at: Optional[datetime] = None
    error: Optional[str] = None
    artifacts: List[ArtifactOut] = []

    model_config = {"from_attributes": True}


BatchDetailOut.model_rebuild()

File: frontend/src/api/subscription.ts
import type { AxiosResponse } from 'axios';
import api from './client';

// 邮局订报数据生成模块 —— 上传驱动、版本流水、多文件生成。

export type BatchStatus = 'draft' | 'pending_validation' | 'ready' | 'generated' | 'archived';
export type ImportStatus =
  | 'uploading' | 'parsing' | 'validation_failed' | 'validation_passed' | 'active' | 'superseded';
export type IssueLevel = 'block' | 'warn' | 'info';
export type RunStatus = 'queued' | 'running' | 'success' | 'failed' | 'void';
export type ArtifactType = 'workbook' | 'postal_summary' | 'region_detail' | 'zip';

export interface SubBatch {
  id: number;
  year: number;
  start_month: number;
  make_date: string | null;
  unit_price: string | null;
  status: BatchStatus;
  active_version_id: number | null;
  notes: string | null;
  created_at: string | null;
  updated_at: string | null;
}

export interface SourceFile {
  id: number;
  file_role: string;
  file_type: string | null;
  original_filename: string;
  size: number | null;
  sha256: string | null;
}

export interface ImportVersion {
  id: number;
  batch_id: number;
  version_no: number;
  status: ImportStatus;
  reason: string | null;
  summary_json: Record<string, unknown> | null;
  uploaded_at: string | null;
  source_files: SourceFile[];
}

export interface BatchDetail extends SubBatch {
  versions: ImportVersion[];
}

export interface ValidationIssue {
  id: number;
  level: IssueLevel;
  source: string | null;
  sheet_or_file: string | null;
  row_no: number | null;
  field: string | null;
  code: string | null;
  message: string;
}

export interface SubRecord {
  id: number;
  name: string;
  phone: string | null;
  province: string | null;
  city: string | null;
  district: string | null;
  address: string | null;
  postal_code: string | null;
  copies: number;
  months: number | null;
  amount: string | null;
  region_name: string | null;
  source_channel: string | null;
  remittance_name: string | null;
  remittance_date: string | null;
  source_file_role: string | null;
  source_row: number | null;
  excluded: boolean;
  exclude_reason: string | null;
}

export interface ImportStatusOut {
  version: ImportVersion;
  issue_counts: Record<IssueLevel, number>;
  can_activate: boolean;
}

export interface Artifact {
  id: number;
  artifact_type: ArtifactType;
  region_name: string | null;
  filename: string;
  sha256: string | null;
  is_historical: boolean;
  created_at: string | null;
}

export interface GenerationRun {
  id: number;
  batch_id: number;
  version_id: number;
  rule_version: string | null;
  template_version: string | null;
  status: RunStatus;
  started_at: string | null;
  ended_at: string | null;
  error: string | null;
  artifacts: Artifact[];
}

export interface BatchCreatePayload {
  year: number;
  start_month: number;
  make_date?: string | null;
  unit_price?: number | null;
  notes?: string | null;
}

export function listSubBatches(): Promise<AxiosResponse<SubBatch[]>> {
  return api.get('/subscription/batches');
}

export function getSubBatch(id: number): Promise<AxiosResponse<BatchDetail>> {
  return api.get(`/subscription/batches/${id}`);
}

export function createSubBatch(body: BatchCreatePayload): Promise<AxiosResponse<SubBatch>> {
  return api.post('/subscription/batches', body);
}

export function createSubImport(
  batchId: number, fileA: File, fileB: File | null, reason?: string,
): Promise<AxiosResponse<ImportVersion>> {
  const fd = new FormData();
  fd.append('file_a', fileA);
  if (fileB) fd.append('file_b', fileB);
  if (reason) fd.append('reason', reason);
  return api.post(`/subscription/batches/${batchId}/imports`, fd);
}

export function getSubImport(versionId: number): Promise<AxiosResponse<ImportStatusOut>> {
  return api.get(`/subscription/imports/${versionId}`);
}

export function getSubImportIssues(versionId: number): Promise<AxiosResponse<ValidationIssue[]>> {
  return api.get(`/subscription/imports/${versionId}/issues`);
}

export function getSubImportRecords(versionId: number): Promise<AxiosResponse<SubRecord[]>> {
  return api.get(`/subscription/imports/${versionId}/records`);
}

export function activateSubImport(versionId: number): Promise<AxiosResponse<ImportVersion>> {
  return api.post(`/subscription/imports/${versionId}/activate`);
}

export function generateSubBatch(batchId: number): Promise<AxiosResponse<GenerationRun>> {
  return api.post(`/subscription/batches/${batchId}/generate`);
}

export function listSubArtifacts(batchId: number): Promise<AxiosResponse<Artifact[]>> {
  return api.get(`/subscription/batches/${batchId}/artifacts`);
}

export async function downloadSubArtifact(id: number, filename: string): Promise<void> {
  const res = await api.get(`/subscription/artifacts/${id}/download`, { responseType: 'blob' });
  const url = URL.createObjectURL(res.data as Blob);
  const a = document.createElement('a');
  a.href = url;
  a.download = filename;
  document.body.appendChild(a);
  a.click();
  a.remove();
  URL.revokeObjectURL(url);
}
